### PR TITLE
8322996: BoxLockNode creation fails with assert(reg < CHUNK_SIZE) failed: sanity

### PR DESCRIPTION
**Pull request by dlunde**
This changeset fixes an issue where deeply nested synchronized statements triggered an assert in C2.

Changes:
- Bail out on compilation when we create a `BoxLockNode` with a slot index that cannot fit in a `RegMask`. This is similar to how we handle the case when we do not have space to represent arguments in [`opto/matcher.cpp`](https://github.com/openjdk/jdk/blob/58b01dce054c50bcb5a28aad4c1b574acaa90f6d/src/hotspot/share/opto/matcher.cpp#L314-L318)
- Generalize `RegMask::can_represent` to take an additional and optional size argument to facilitate reuse. The default size value, 1, corresponds to the previous functionality. Rewrite `can_represent_arg` to directly call `can_represent(reg, SlotsPerVecZ)`.
- Add a regression test.

Testing:
- [GitHub Actions](https://github.com/dlunde/jdk/actions/runs/7671871191)
- tier1 to tier5 on windows-x64, linux-x64, linux-aarch64, macosx-x64, and macosx-aarch64.
- The new regression test in all tier1 to tier10 contexts on windows-x64, linux-x64, linux-aarch64, macosx-x64, and macosx-aarch64.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8322996](https://bugs.openjdk.org/browse/JDK-8322996): BoxLockNode creation fails with assert(reg &lt; CHUNK_SIZE) failed: sanity (**Bug** - P3)

### Reviewers
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**) ⚠️ Review applies to [524438ca](https://git.openjdk.org/jdk/pull/17370/files/524438caa251e9816232f8ce6fd92304d8bdaf65)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [ebe23a23](https://git.openjdk.org/jdk/pull/17370/files/ebe23a233d40da788835d8506f7500db89b3e16c)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17370/head:pull/17370` \
`$ git checkout pull/17370`

Update a local copy of the PR: \
`$ git checkout pull/17370` \
`$ git pull https://git.openjdk.org/jdk.git pull/17370/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17370`

View PR using the GUI difftool: \
`$ git pr show -t 17370`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17370.diff">https://git.openjdk.org/jdk/pull/17370.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17370#issuecomment-1886801661)

### CODE REVIEW

File: src/hotspot/share/opto/locknode.cpp [hunk 1]
```diff
@@ -44,6 +44,10 @@ BoxLockNode::BoxLockNode( int slot ) : Node( Compile::current()->root() ),
   init_class_id(Class_BoxLock);
   init_flags(Flag_rematerialize);
   OptoReg::Name reg = OptoReg::stack2reg(_slot);
+  if (!RegMask::can_represent_arg(reg)) {
```
Comment (robcasloz) [inline src/hotspot/share/opto/locknode.cpp:47]: I am not very familiar with this code, but would it be possible to use `!RegMask::can_represent(reg)` instead of `!RegMask::can_represent_arg(reg)` here? Or is it necessary to use the latter (which is stricter) for correctness?

File: test/hotspot/jtreg/compiler/c2/TestNestedSynchronize.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @bug 8322996
+ * @requires vm.debug
```
Comment (robcasloz) [inline test/hotspot/jtreg/compiler/c2/TestNestedSynchronize.java:27]: I suggest removing this line for better test coverage, the test does not really require debug mode.

File: test/hotspot/jtreg/compiler/c2/TestNestedSynchronize.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+ * @summary Ensure no assert error in C2 with deeply nested synchronize
+ *          statements.
+ *
+ * @run main/othervm -XX:CompileCommand=compileonly,compiler.c2.TestNestedSynchronize::test
+ *                   -XX:-TieredCompilation -Xcomp
```
Comment (robcasloz) [inline test/hotspot/jtreg/compiler/c2/TestNestedSynchronize.java:32]: No need to use `-XX:-TieredCompilation` here (already in `-Xcomp` mode).

File: test/hotspot/jtreg/compiler/c2/TestNestedSynchronize.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+ *                   compiler.c2.TestNestedSynchronize
+ */
+
+package compiler.c2;
```
Comment (robcasloz) [inline test/hotspot/jtreg/compiler/c2/TestNestedSynchronize.java:36]: The test case might fit better under `test/hotspot/jtreg/compiler/locks`.

**Review by robcasloz — CHANGES_REQUESTED** [review]

File: src/hotspot/share/opto/locknode.cpp (hunk 1 above)
Comment (dlunde) [reply src/hotspot/share/opto/locknode.cpp:47]: That is a fair question, and I'm not sure what is the preferred solution.

The number of stack slots for a monitor seems to be determined by [`sync_stack_slots`](https://github.com/dlunde/jdk/blob/06d6b4be9750a326f87acf04a3dc717e307d14d5/src/hotspot/share/opto/compile.hpp#L1166-L1167). If I'm not mistaken the value of `sync_stack_slots()` varies between platforms (on my machine it is `int Compile::sync_stack_slots() const { return 2; }`). Therefore, I don't think `can_represent` always works. However, we should maybe have a new function `can_represent_sync_entry` (or similar) instead of `can_represent_arg`.

File: src/hotspot/share/opto/locknode.hpp [hunk 5]
```diff
@@ -57,6 +57,13 @@ class BoxLockNode : public Node {
   }
   int stack_slot() const { return _slot; }
 
+  // Check if a RegMask is big enough to represent a synchronization entry at
+  // the index reg. Analogous to can_represent and can_represent_arg in
+  // regmask.hpp.
+  static bool can_represent_sync(OptoReg::Name reg) {
+    return (int)reg < (int)(RegMask::CHUNK_SIZE - 1 - Compile::current()->sync_stack_slots());
+  }
+
```
Comment (vnkozlov) [inline src/hotspot/share/opto/locknode.hpp:66]: I think it should be in `regmask.hpp` together with other `can_represent_*` methods. Then you don't need part of the comment about those methods.

Comment (dlunde) [reply src/hotspot/share/opto/locknode.hpp:66]: Thanks @vnkozlov. Do you know if we can directly use `can_represent` instead, and not take `sync_stack_slots()` into account? The field `_inmask` in `BoxLockNode` seems to only specify a single register (one bit in the mask).

Comment (vnkozlov) [reply src/hotspot/share/opto/locknode.hpp:66]: I think your current code is correct.

On x64 `sync_stack_slots` defined as 2 (takes 2 bits in regmask) in  `x86_64.ad` and as 1 in `x86_32.ad`. On most 64 bit platforms it is also 2 slots, from what I see. But we can't guarantee that some platforms will not have bigger value. We can't use last odd bit on 64 bit platform in regmask - it is taking anyway already by "infinite stack flag".

Comment (dlunde) [reply src/hotspot/share/opto/locknode.hpp:66]: Yes, that is my intuition as well. Therefore, I'm left wondering if the [construction of `_inmask`](https://github.com/dlunde/jdk/blob/9ab6e561780aee0f2cc2f06cd40ec487d60fe39c/src/hotspot/share/opto/locknode.cpp#L51) in the `BoxLockNode` constructor is incorrect, as it always just sets a single bit in the mask (no matter the value of `sync_stack_slots()`). Should we perhaps change it to instead set the range [ reg, reg + sync_stack_slots() ) in `_inmask`?

Comment (vnkozlov) [reply src/hotspot/share/opto/locknode.hpp:66]: It is fine. Monitors use fixed stack slots which are not available to Register allocator: [compile.hpp#L315](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/compile.hpp#L315). Fixed stack slots reserved based on number monitors and etc. We don't do regmask operations on them.

`BoxLockNode::reg(box)` is only accessed in format output in `JVMState::format()` and when we generate debug info in `PhaseOutput::Process_OopMap_Node()`.

In general, in 64-bit VM long, double, oop/pointers values take 2 slots but their value is stored only in first slot (since slot's size is 64 bit). Don't ask why. It is convention in VM.

**Comment by dlunde** [comment]
@robcasloz  @vnkozlov: I have now made the changes we've discussed. Please have a look when you have some time to spare.

File: test/hotspot/jtreg/compiler/locks/TestNestedSynchronize.java [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+ * @test
+ * @bug 8322996
+ * @summary Ensure no assert error in C2 with deeply nested synchronize
+ *          statements.
+ * @run main/othervm -XX:CompileCommand=compileonly,compiler.c2.TestNestedSynchronize::test
+ *                   -Xcomp
+ *                   compiler.c2.TestNestedSynchronize
+ */
+
+package compiler.c2;
```
Comment (robcasloz) [inline test/hotspot/jtreg/compiler/locks/TestNestedSynchronize.java:34]: Update package name to `compiler.locks`.

**Review by robcasloz — CHANGES_REQUESTED** [review]
The fix itself looks good to me.
Would it make sense, for better coverage, to add a couple of additional test cases that exercise the boundaries of the condition that is tested? E.g. one with one `synchronized` statement less than the current one and one with one  `synchronized` statement more.

File: test/hotspot/jtreg/compiler/locks/TestNestedSynchronize.java (hunk 6 above)
Comment (dlunde) [reply test/hotspot/jtreg/compiler/locks/TestNestedSynchronize.java:34]: Oops, thanks. Updating and rerunning tests.

**Comment by dlunde** [comment]
> The fix itself looks good to me. Would it make sense, for better coverage, to add a couple of additional test cases that exercise the boundaries of the condition that is tested? E.g. one with one `synchronized` statement less than the current one and one with one `synchronized` statement more.

I have experimented with such test cases (various edge cases) and as a result found a related (but separate) issue from this one. I was planning to add these additional tests for that separate issue, to not introduce unnecessary test failures before that fix is integrated. Maybe it is better to add the additional tests directly as part of this changeset instead?

**Comment by robcasloz** [comment]
> I have experimented with such test cases (various edge cases) and as a result found a related (but separate) issue from this one. I was planning to add these additional tests for that separate issue, to not introduce unnecessary test failures before that fix is integrated. Maybe it is better to add the additional tests directly as part of this changeset instead?

If the additional tests trigger failures after this fix is applied, I would suggest including them as part of the fix to the separate issue.

**Review by robcasloz — APPROVED** [review]

File: src/hotspot/share/opto/graphKit.cpp [hunk 7]
```diff
@@ -3469,6 +3469,8 @@ FastLockNode* GraphKit::shared_lock(Node* obj) {
 
   // Box the stack location
   Node* box = _gvn.transform(new BoxLockNode(next_monitor()));
+  // Check for bailout after new BoxLockNode
+  if (failing()) { return nullptr; }
```
Comment (vnkozlov) [inline src/hotspot/share/opto/graphKit.cpp:3473]: Do all callers of `shared_lock()` checks for `failing()` or returned `nullptr`?

**Review by vnkozlov — COMMENTED** [review]
Looks good. I have one question.

File: src/hotspot/share/opto/graphKit.cpp (hunk 7 above)
Comment (dlunde) [reply src/hotspot/share/opto/graphKit.cpp:3473]: No, not the immediate callers at least. Below is a quick call graph analysis for the places where we create `BoxLockNode`s (up to the first bailout check). Should I add returns at all points in the call chain up to the first checks?
```
graphKit.cpp:3471 (this is in shared_lock)
  locknode.cpp:196
    parse2.cpp:2759
      parse1.cpp:1594 (Checks for bailout at parse1.cpp:1595)
  parse1.cpp:1264
    parse1.cpp:582 (Checks for bailout at parse1.cpp:596)

parse1.cpp:227
  parse1.cpp:579 (Checks for bailout at parse1.cpp:596)
```

Comment (vnkozlov) [reply src/hotspot/share/opto/graphKit.cpp:3473]: I am not worry about exit in`load_interpreter_state()` there is check after call.

There is call to `record_profiled_parameters_for_speculation()` after `shared_lock()` call in `do_method_entry()`. It does not check `failing()`. May be add check `if (!failing())` when calling it (or inside it).

For `do_one_bytecode()` case the only question is if IGVN code can handle such bailout. It is called at the end of method. If not, we need to add the check.

Comment (dlunde) [reply src/hotspot/share/opto/graphKit.cpp:3473]: > There is call to record_profiled_parameters_for_speculation() after shared_lock() call in do_method_entry(). It does not check failing(). May be add check if (!failing()) when calling it (or inside it).

I've now added a check just after `_synch_lock = shared_lock(lock_obj);`, thanks.

> For do_one_bytecode() case the only question is if IGVN code can handle such bailout. **It is called at the end of method**. If not, we need to add the check.

I cannot see any call to IGVN after `do_monitor_enter()` in `do_one_bytecode()`. Can you elaborate? Maybe related, I also changed to
```
Node *box = new BoxLockNode(next_monitor());
// Check for bailout after new BoxLockNode
if (failing()) { return; }
box = _gvn.transform(box);
```
from the previous
```
Node *box = _gvn.transform(new BoxLockNode(next_monitor()));
// Check for bailout after new BoxLockNode
if (failing()) { return; }
```
to avoid potential issues with `_gvn.transform` due to the new bailout.

Comment (vnkozlov) [reply src/hotspot/share/opto/graphKit.cpp:3473]: Does `_gvn.transform` handle bailout?

> I cannot see any call to IGVN after `do_monitor_enter()` in `do_one_bytecode()`. Can you elaborate?
Sorry, I mean IGV (IdealGraphVisualizer) [parse2.cpp#L2782](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/parse2.cpp#L2782)

**Review by vnkozlov — APPROVED** [review]
Looks good.
